kernel: Define application entry point prototype

aescolar opened the pull request:
kernel: Define application entry point prototype
    
Define in a public header the prototype of the (optional) application `main()`.
    
Until now, this was hidden inside the init code, which lead to some tests and samples defining it with the wrong signature.
    
Also this fixes issues when defining this function in C++ code.

~Note that this PR includes #21100, as that fix is necessary for it~
~But I would not merge this PR until after freeze (as it may break apps which have the same issue as fixed in #21100)~

~Fixes #21094~(already fixed)

aescolar committed:
kernel: Define application entry point prototype

Define in a public header the prototype of the (optional) application
main().

Until now, this was hidden inside the init code, which lead to
some tests and samples defining it with the wrong signature.

Also this fixes issues when defining this function in C++ code.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

pabigot reviewed (approved):
Approval specific to last commit, rest covered by #21100.

pfalcon reviewed (changes requested):
Per my comment in https://github.com/zephyrproject-rtos/zephyr/pull/21100#pullrequestreview-325259287 , I'd suggest this to be `int main(void)`, to follow the principle of the least surprise (and let the multitude of `int main(void)` around be the extra proof that it's indeed the least surprise.)

aescolar committed:
ztest: Separate unit test main into its own file

Zephyr headers tend to pull lots of other headers which
tend to create dependency loops and complicate treating
some parts of the code differently.
As the unit testing main() is actually a plain C main(),
meant to be compiled as a host binary, it needs some
special treatment compared to other Zephyr files.

So let's split it into its own file with minimal dependencies.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

dbkinder commented on include/zephyr.h:
```diff
@@ -17,4 +17,23 @@
+ * be called by the kernel after it completes the initialization.
+ * Such a function must take no arguments and return no value.
+ *
+ * This function may return but does not have to.
```
Do we define what happens If the main function does return, e.g., the system will shutdown/power off?

andrewboie replied:
There's no special semantics currently, it will just cause the main thread to exit.

aescolar replied:
@dbkinder Would you like me to either link to that [part of the documentation](https://docs.zephyrproject.org/latest/reference/kernel/threads/system_threads.html) or replicate in this comment part of it? (It's a bit unclear to me how you would like to proceed)

dbkinder replied:
@andrewboie @aescolar @nashif  I guess I was looking for a defined behavior of the system if the main function returns and that thread is terminated. Architecturally, what code does the processor execute after that then and what will the system be doing?  If that's not defined by Zephyr, then so be it.

andrewboie replied:
another ready thread will then run, which may be the idle thread if no other threads are present or ready to run.

main is just another thread. nothing special happens when it returns.

andrewboie replied:
if all threads exit then the system just sits there in the idle thread for each cpu.

dbkinder replied:
Ah.  Thanks for clarifying that for me @andrewboie, maybe it's obvious that there's always an idle thread around for a CPU to run.

dbkinder commented:
We should also update https://docs.zephyrproject.org/latest/reference/kernel/threads/system_threads.html?highlight=main%20function to indicate what happens when ``main()`` returns.

aescolar commented:
> We should also update [..] to indicate what happens when main() returns.

@dbkinder It is already specified there what happens:
`If main() is not defined, or if it executes and then does a normal return, the main thread terminates normally and no error is raised.`

ceolin commented on subsys/testsuite/ztest/include/ztest_main_if.h:
```diff
@@ -0,0 +1,22 @@
+ */
+
+#ifndef SUBSYS_TESTSUITE_ZTEST_INCLUDE_ZTEST_MAIN_IF_H
+#define SUBSYS_TESTSUITE_ZTEST_INCLUDE_ZTEST_MAIN_IF_H
```
Why don't you name it `ztest_unit.h` or `ztest_unit_main.h` ? It seems more coherent.

ceolin replied:
Actually I'm not getting the purpose of this commit, even for unit test one still need to include `ztest.h` because there may be some required definitions there, right ? Even if not, you are still building `ztest.c` for unit tests and `ztest.h` is still including this header  ....

It seems to me that this is just bringing more complication ... but I may be wrong here.

aescolar replied:
The purpose of this, is being able to compile ztest_unit_main.c without pulling all other zephyr headers into that translation unit. 
This ztest_unit_main defines the Linux executable `main()`. Meaning, it is just part of the "unit tests framework".
Compiling it with other zephyr headers as it was, complicates things quite a lot unnecessarily. Separating it, is the easiest way of having the new definition this PR introduces not create problems in the unit test framework. It should also avoid other intermingling problems in the future.
Wrt. this new header in particular, it is just what this `main()` requires from the rest of the unit test framework.

ceolin commented on samples/cpp_synchronization/src/main.cpp:
```diff
@@ -155,5 +155,5 @@ int main(void)
 	}
 
-	return 0;
+	return;
```
not necessary return in these cases, just remove them. Lets be concise with all others `void` functions in our code base.

ceolin reviewed (changes requested):
Just small things to take a look, otherwise it is ok to me.

aescolar commented:
#21094 was fixed by #21363, so as such there is no special need to merge this one. Given that there is some controversy on this one I close it, but leave the branch in case somebody wants to take it further.
